agent/engine/dockerstate/docker_task_engine_state.go

```diff
@@ -364,18 +364,27 @@ func (state *DockerTaskEngineState) RemoveTask(task *api.Task) {
 
 	task, ok := state.tasks[task.Arn]
 	if !ok {
+		seelog.Warnf("Failed to locate task %s for removal from state", task.Arn)
 		return
 	}
 	delete(state.tasks, task.Arn)
+
 	containerMap, ok := state.taskToID[task.Arn]
 	if !ok {
+		seelog.Warnf("Failed to locate containerMap for task %s for removal from state", task.Arn)
 		return
 	}
 	delete(state.taskToID, task.Arn)
 
 	for _, dockerContainer := range containerMap {
-		delete(state.idToTask, dockerContainer.DockerID)
-		delete(state.idToContainer, dockerContainer.DockerID)
+		// The key to these maps is either the Docker ID or agent-generated name.  We use the agent-generated name
+		// before a Docker ID is available.
+		key := dockerContainer.DockerID
```

Comment:
it might be safer to move this out of this method into a `getKey(container)` method (so that future modifications in this method do not touch that critical piece of business logic). what do you think?

Reply:
The code that adds to the map also removes references during a subsequent call.  I'll extract both operations out to explicit functions with comments explaining why they're doing this.